### ISSUE TITLE
Support for server-side rendering via a Babel plugin?

### ISSUE THREAD

**Issue by brandonweiss** [mdx-js/mdx#171]
This project is super cool! I was just trying to figure out the best way to add markdown support to the JSX-based static site generator I’m working on and then I found this—it’s exactly what I’m looking for!

I’m struggling to figure out how to use it, though. I see that there is a Webpack loader, but I’m not using Webpack—the files aren’t being bundled for the browser. I see there is also an API for transforming MDX strings/files, but the problem is I need to actually require/run the JavaScript locally, because the JSX/MDX is being rendered server-side by React.

I think the general thing people do is hook into Node’s `require` using `babel-register`.

```node
babelRegister({
  extensions: [".jsx"],
  presets: [
    "react",
    [
      "env",
      {
        node: "current",
      }
    ]
  ],
})
```

Then you can `require` the JSX file/module and render it. If the module requires/imports other files they’ll also get transpiled because of `babel-register`.

```node
let component = require(modulePath)
component = component.default || component

let element = React.createElement(component)

return ReactDOMServer.renderToStaticMarkup(element)
```

I found a relatively old project [babel-plugin-transform-mdx](https://github.com/9dots/babel-plugin-transform-mdx). I tried using it but it seems to be broken.

```shell
❯ cat test.mdx 
# Hello, world!

❯ babel --plugins transform-mdx test.mdx
SyntaxError: test.mdx: Unexpected character '#' (1:0)
> 1 | # Hello, world!
    | ^
  2 | 

```

I’ve got some experience with ASTs so I’d normally get my hands dirty and figure out how to make this work, but after reading the MDX transform in that repo and the JSX transform included with Babel, I’m finding the barrier to entry to be extremely high. 😞

Some time ago JSX became ubiquitous enough that it seemingly made sense for Babel to take ownership of JSX transformation. Since MDX is dependent on (or a sibling of) JSX, it makes sense to me that it should be usable in the same way that JSX is. Does the MDX team agree? Would you be willing to add support for a Babel plugin? 🙏

**Comment by timneutkens** [comment]
> I found a relatively old project babel-plugin-transform-mdx. I tried using it but it seems to be broken.

This is actually unrelated to `mdx-js/mdx` it's not focused on the implementation/specification we wrote for MDX.

**Comment by timneutkens** [comment]
It's probably good to read this comment I wrote about how MDX works: https://github.com/mdx-js/mdx/issues/126#issuecomment-389073920

Basically the core mdx package `@mdx-js/mdx` provides methods to turn an `mdx` string into an es2015 module that can be parsed by any es2015 parser, usually this is done by Babel.

**Comment by brandonweiss** [comment]
Hmm, I think I already understood that… I know that MDX transforms to an ES6 module with JSX. But I’m not quite clear of what the implication of what you’re saying is regarding what I want to do… can you dumb it down for me a bit? 😅

Are you saying that I want to do is not possible? Or easier than I think it is? Or just not the responsibility of this project?

**Comment by timneutkens** [comment]
Looks like it uses this: https://github.com/ariporad/pirates

**Comment by brandonweiss** [comment]
Ohhhh, now I understand. Thanks for explaining!

Yeah! I was spelunking in the `babel-register` sub-package earlier today to figure out how they overrode `require` and found the pirates package. I think I know what to do now. 👌

**Comment by timneutkens** [comment]
Feel free to share the solution you figure out here 🙏

**Comment by brandonweiss** [comment]
Alright, here’s what I wound up doing.

The first part is setting up the transformation hook.

```javascript
import { addHook as overrideRequire } from "pirates"
import { sync as mdxTransform } from "@mdx-js/mdx"
import { transform as babelTransform } from "babel-core"

const transform = (code) => {
  let jsxWithMDXTags = mdxTransform(code)

  let jsx = `
    import { MDXTag } from "@mdx-js/tag"

    ${jsxWithMDXTags}
  `

  let result = babelTransform(jsx, {
    plugins: "babel-plugin-react-require",
    presets: [
      "babel-preset-react",
      [
        "babel-preset-env",
        {
          node: "9.7.1",
        },
      ],
    ],
  })

  return result.code
}

overrideRequire(transform, { exts: [".mdx"] })
```

The second part is actually importing/requiring the file which triggers the transformation.

```javascript
import React from "react"
import ReactDOMServer from "react-dom/server"

export default async (modulePath, props) => {
  let component = await import(modulePath)
  let element = React.createElement(component.default, props)
  let html = ReactDOMServer.renderToStaticMarkup(element)

  return `<!DOCTYPE html>\n\n${html}`
}
```

This isn't the _exact_ code—I’ve simplified it to make it easier to understand, but this is the general idea. Also this uses ES6. You'll need to convert to CommonJS or use [`esm`](https://github.com/standard-things/esm).
